# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1284996062 14400
#      Mon Sep 20 11:21:02 2010 -0400
# Node ID 2538d03489cc4e7554858f44891e9c165d29b74d
# Parent  2ab8cc6dccedd2e3a3c7b3a4ea9e79ce2a884a2c
avoid infinite loop in regexp searches with PCRE

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,14 @@
+2010-09-20  John W. Eaton  <jwe@octave.org>
+
+	* DLD-FUNCTIONS/regexp.cc (octregexp_list) [HAVE_PCRE]:
+	Break from search loop if idx is at or beyond end of buffer
+	(bug #31056).
+
 2010-09-19  Ben Abbott <bpabbott@mac.com>
 
 	* graphics.h.in: Add displayname property to patch and surface objects.
 
 2010-09-18  David Bateman  <dbateman@free.fr>
 
 	* graphics.h.in: Remove properties for dead legend code.
 
diff --git a/src/DLD-FUNCTIONS/regexp.cc b/src/DLD-FUNCTIONS/regexp.cc
--- a/src/DLD-FUNCTIONS/regexp.cc
+++ b/src/DLD-FUNCTIONS/regexp.cc
@@ -481,17 +481,17 @@ octregexp_list (const octave_value_list 
 
               pcre_free_substring_list(listptr);
 
               regexp_elem new_elem (named_tokens, t, m, te, s, e);
               lst.push_back (new_elem);
               idx = ovector[1];
               sz++;
 
-              if (once)
+              if (once || idx >= buffer.length ())
                 break;
 
             }
         }
 
       pcre_free(re);
 #else
       regex_t compiled;
